File: app/src/main/java/it/liceoarzignano/bold/news/NewsHandler.java
package it.liceoarzignano.bold.news;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import it.liceoarzignano.bold.database.DBHandler;
import it.liceoarzignano.bold.utils.DateUtils;

public class NewsHandler extends DBHandler<News> {
    private static final String DB_NAME = "NewsDatabase.db";
    private static final int DB_VERSION = 1;
    private static final String KEY_TITLE = "title";
    private static final String KEY_DATE = "date";
    private static final String KEY_DESCRIPTION = "description";
    private static final String KEY_URL = "url";

    private static NewsHandler sInstance;

    public static synchronized NewsHandler getInstance(Context context) {
        if (sInstance == null) {
            sInstance = new NewsHandler(context);
        }
        return sInstance;
    }

    private NewsHandler(Context context) {
        super(context, DB_NAME, DB_VERSION);
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        db.execSQL("CREATE TABLE " + getTableName() + " (" +
                KEY_ID + " INTEGER PRIMARY KEY, " +
                KEY_TITLE + " TEXT, " +
                KEY_DATE + " INTEGER, " +
                KEY_DESCRIPTION + " TEXT, " +
                KEY_URL + " TEXT)");
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        // Update this when db table will be changed
    }

    @Override
    public List<News> getAll() {
        List<News> list = new ArrayList<>();
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery("SELECT * FROM " + getTableName() +
                " ORDER BY " + KEY_DATE + " DESC", null);

        if (cursor.moveToFirst()) {
            do {
                list.add(new News(
                        Long.parseLong(cursor.getString(0)),
                        cursor.getString(1),
                        Long.parseLong(cursor.getString(2)),
                        cursor.getString(3),
                        cursor.getString(4)
                ));
            } while (cursor.moveToNext());
        }

        cursor.close();
        return list;
    }

    @Override
    @Nullable
    public News get(long id) {
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery("SELECT * FROM " + getTableName() +
                " WHERE " + KEY_ID + "=?", new String[] { String.valueOf(id) });

        News news = null;
        if (cursor.moveToFirst()) {
            news = new News(
                    Long.parseLong(cursor.getString(0)),
                    cursor.getString(1),
                    Long.parseLong(cursor.getString(2)),
                    cursor.getString(3),
                    cursor.getString(4)
            );
        }
        cursor.close();
        return news;
    }

    @Override
    protected ContentValues getValues(@NonNull News item, boolean withId) {
        ContentValues values = new ContentValues();
        values.put(KEY_TITLE, item.getTitle());
        values.put(KEY_DATE, item.getDate());
        values.put(KEY_DESCRIPTION, item.getDescription());
        values.put(KEY_URL, item.getUrl());

        if (withId) {
            values.put(KEY_ID, item.getId());
        }

        return values;
    }

    @Override
    protected String getTableName() {
        return "news";
    }

    List<News> getByQuery(@Nullable String query) {
        List<News> list = getAll();

        if (query != null) {
            list.removeIf(it -> {
                String title = it.getTitle().toLowerCase();
                String message = it.getDescription().toLowerCase();
                String date = DateUtils.dateToString(new Date(it.getDate())).toLowerCase();
                return !title.contains(query.toLowerCase()) &&
                        !date.contains(query.toLowerCase()) &&
                        !message.contains(query.toLowerCase());
            });
        }

        return list;
    }
}

File: app/src/main/kotlin/it/liceoarzignano/bold/news/NewsHandler.kt
package it.liceoarzignano.bold.news

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import it.liceoarzignano.bold.database.DBHandler

class NewsHandler private constructor(context: Context) : DBHandler<News>(context, DB_NAME, DB_VERSION) {

    override fun onCreate(db: SQLiteDatabase) = db.execSQL("CREATE TABLE $tableName (" +
            "$KEY_ID INTEGER PRIMARY KEY AUTOINCREMENT, " +
            "$KEY_TITLE TEXT, " +
            "$KEY_DATE INTEGER, " +
            "$KEY_DESCRIPTION TEXT, " +
            "$KEY_URL TEXT," +
            "$KEY_UNREAD INTEGER)")

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        if (oldVersion == 1 && newVersion == 2) {
            db.execSQL("ALTER TABLE $tableName ADD COLUMN $KEY_UNREAD INTEGER DEFAULT 0")
        }
    }

    public override val all: MutableList<News>
        get() {
            val list = ArrayList<News>()
            val db = readableDatabase
            val cursor = db.rawQuery("SELECT * FROM $tableName ORDER BY $KEY_DATE DESC", null)

            if (cursor.moveToFirst()) {
                do {
                    list.add(News(
                            cursor.getLong(0),
                            cursor.getString(1),
                            cursor.getLong(2),
                            cursor.getString(3),
                            cursor.getString(4),
                            cursor.getInt(5) == 1))
                } while (cursor.moveToNext())
            }

            cursor.close()
            return list
        }

    override fun get(id: Long): News? {
        val db = readableDatabase
        val cursor = db.rawQuery("SELECT * FROM $tableName WHERE ${DBHandler.KEY_ID} = ?",
                arrayOf(id.toString()))

        var news: News? = null
        if (cursor.moveToFirst()) {
            news = News(
                    cursor.getLong(0),
                    cursor.getString(1),
                    cursor.getLong(2),
                    cursor.getString(3),
                    cursor.getString(4),
                    cursor.getInt(5) == 1)
        }
        cursor.close()
        return news
    }

    override fun getValues(item: News): ContentValues {
        val values = ContentValues()
        values.put(KEY_TITLE, item.title)
        values.put(KEY_DATE, item.date)
        values.put(KEY_DESCRIPTION, item.description)
        values.put(KEY_URL, item.url)
        values.put(KEY_UNREAD, if (item.unread) 1 else 0)
        return values
    }

    override val tableName: String get() = "news"

    fun getByQuery(query: String?): List<News> {
        if (query == null || query.isBlank()) {
            return all
        }

        val list = ArrayList<News>()
        val db = readableDatabase
        val cursor = db.rawQuery("SELECT * FROM $tableName WHERE $KEY_TITLE LIKE ?",
                arrayOf("%$query%"))
        if (cursor.moveToFirst()) {
            do {
                list.add(News(
                        cursor.getLong(0),
                        cursor.getString(1),
                        cursor.getLong(2),
                        cursor.getString(3),
                        cursor.getString(4),
                        cursor.getInt(5) == 1))
            } while (cursor.moveToNext())
        }

        cursor.close()
        return list
    }

    fun getByUrl(query: String): List<News> {
        val list = ArrayList<News>()
        val db = readableDatabase
        val cursor = db.rawQuery("SELECT * FROM $tableName WHERE $KEY_URL LIKE ?",
                arrayOf("%$query%"))
        if (cursor.moveToFirst()) {
            do {
                list.add(News(
                        cursor.getLong(0),
                        cursor.getString(1),
                        cursor.getLong(2),
                        cursor.getString(3),
                        cursor.getString(4),
                        cursor.getInt(5) == 1))
            } while (cursor.moveToNext())
        }

        cursor.close()
        return list

    }

    fun getUnread(): List<News> {
        val list = ArrayList<News>()
        val db = readableDatabase
        val cursor = db.rawQuery("SELECT * FROM $tableName ORDER BY $KEY_DATE DESC, $KEY_UNREAD DESC", emptyArray())
        if (cursor.moveToFirst()) {
            do {
                list.add(News(
                        cursor.getLong(0),
                        cursor.getString(1),
                        cursor.getLong(2),
                        cursor.getString(3),
                        cursor.getString(4),
                        cursor.getInt(5) == 1))
            } while (cursor.moveToNext())
        }

        cursor.close()
        return list
    }

    fun hasUnread(): Boolean {
        val db = readableDatabase
        val cursor = db.rawQuery("SELECT * FROM $tableName WHERE $KEY_UNREAD = 1", emptyArray())

        val hasItems = cursor.count != 0
        cursor.close()
        return hasItems
    }

    companion object {
        private val DB_NAME = "NewsDatabase.db"
        private val DB_VERSION = 2
        private val KEY_TITLE = "title"
        private val KEY_DATE = "date"
        private val KEY_DESCRIPTION = "description"
        private val KEY_URL = "url"
        private val KEY_UNREAD = "unread"

        // Singleton
        @Volatile private var INSTANCE: NewsHandler? = null
        fun getInstance(context: Context): NewsHandler =
                INSTANCE ?: synchronized(this) {
                    INSTANCE ?: NewsHandler(context).also { INSTANCE = it }
                }
    }
}
